File: EmojibleIOS/EmojibleIOS/DataSources/TutorialDataSource.swift
//
//  TutorialDataSource.swift
//  EmojibleIOS
//
//  Created by Furkan Yakal on 5.11.2020.
//

import Foundation
import Firebase

class TutorialDataSource {
    
    let database = Firestore.firestore()
    var snapshotListener:ListenerRegistration?
    let notificationCenter = NotificationCenter.default
    var tutorialDataSourceIndices:[String] = []
    
    func startObservingDefaultTutorials(){
        let tutorialsCollectionRef = database.collection("DefaultTutorials")
        snapshotListener = tutorialsCollectionRef.addSnapshotListener{ [unowned self] querySnapshot, error in
            guard let documents = querySnapshot?.documents else { return }
            let defaultTutorials = documents.map{CodeModel(dictionary: $0.data())}
            tutorialDataSourceIndices = documents.map{$0.documentID}
            notificationCenter.post(name: .defaultTutorialsChanged, object: nil, userInfo:["defaultTutorials":defaultTutorials])
        }
    }
    
    func stopObservingDefaultTutorials(){
        guard let snapshotListener = snapshotListener else { return }
        snapshotListener.remove()
    }
    
    func editTutorial(index:Int, newModel: CodeModel){
        database.collection("DefaultTutorials").document(tutorialDataSourceIndices[index]).setData(newModel.dictionary)
    }
    
    private init(){
    }
    private static let instance = TutorialDataSource()
    public static func getInstance() -> TutorialDataSource{
        return .instance
    }
}

File: EmojibleAndroid/app/src/main/java/com/dji/emojibleandroid/dataSources/TutorialDataSource.kt
package com.dji.emojibleandroid.dataSources

import com.dji.emojibleandroid.models.CodeModel
import com.dji.emojibleandroid.services.AuthenticationManager
import com.dji.emojibleandroid.services.Changes
import com.dji.emojibleandroid.services.NotificationCenter
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase

class TutorialDataSource {
    var database = Firebase.firestore
    var snapShotListener: ListenerRegistration? = null
    val notificationCenter: NotificationCenter = NotificationCenter.instance
    var tutorialDataSourceIndices: List<String> = listOf()

    fun startObservingProgram() {
        val tutorialsCollectionRef = database.collection("DefaultTutorials")
        snapShotListener = tutorialsCollectionRef.addSnapshotListener { querySnapshot, _ ->
            val documents = querySnapshot?.documents
            val defaultTutorials = documents?.map { CodeModel(it.data as HashMap<String, Any>) }
            tutorialDataSourceIndices = documents?.map { it.id }!!
            notificationCenter.post(
                Changes.defaultTutorialsChanged,
                null,
                "defaultTutorials" to defaultTutorials
            )
        }
    }

    fun stopObservingProgram() {
        snapShotListener?.remove()
    }

    fun editTutorial(index: Int, newModel: CodeModel) {
        database.collection("DefaultTutorials").document(tutorialDataSourceIndices[index])
            .set(newModel.dictionary)
    }

    private object HOLDER {
        val INSTANCE = TutorialDataSource()
    }

    companion object {
        val instance: TutorialDataSource by lazy { HOLDER.INSTANCE }
    }
}